pkg/webhook/mutate.go

```diff
@@ -31,31 +32,36 @@ func defaultSriovNetworkNodePolicy(cr map[string]interface{}) (*v1beta1.Admissio
 	patchs := []map[string]interface{}{}
 	spec := cr["spec"]
 	if _, ok := spec.(map[string]interface{})["priority"]; !ok {
-		glog.V(2).Infof("defaultSriovNetworkNodePolicy(): set default priority to lowest for %v", name)
+		glog.V(2).Infof("mutateSriovNetworkNodePolicy(): set default priority to lowest for %v", name)
 		patchs = append(patchs, defaultPriorityPatch)
 	}
 	if _, ok := spec.(map[string]interface{})["deviceType"]; !ok {
-		glog.V(2).Infof("defaultSriovNetworkNodePolicy(): set default deviceType to netdevice for %v", name)
+		glog.V(2).Infof("mutateSriovNetworkNodePolicy(): set default deviceType to netdevice for %v", name)
 		patchs = append(patchs, defaultDeviceTypePatch)
 	}
 	if _, ok := spec.(map[string]interface{})["isRdma"]; !ok {
-		glog.V(2).Infof("defaultSriovNetworkNodePolicy(): set default isRdma to false for %v", name)
+		glog.V(2).Infof("mutateSriovNetworkNodePolicy(): set default isRdma to false for %v", name)
 		patchs = append(patchs, defaultDeviceTypePatch)
 	}
 	if _, ok := spec.(map[string]interface{})["linkType"]; !ok {
-		glog.V(2).Infof("defaultSriovNetworkNodePolicy(): set default linkType to eth for %v", name)
+		glog.V(2).Infof("mutateSriovNetworkNodePolicy(): set default linkType to eth for %v", name)
 		patchs = append(patchs, defaultLinkTypePatch)
 	}
-	// MTU is evaluated lastly since default MTU value varies per link type
+	// Default MTU size is evaluated lastly since default MTU value varies per link type
 	if _, ok := spec.(map[string]interface{})["mtu"]; !ok {
 		if str, ok := spec.(map[string]interface{})["linkType"].(string); ok && strings.ToLower(str) == "eth" {
-			glog.V(2).Infof("defaultSriovNetworkNodePolicy(): set default mtu to 1500 for %v", name)
+			glog.V(2).Infof("mutateSriovNetworkNodePolicy(): set default mtu to 1500 for %v", name)
 			patchs = append(patchs, defaultEthMtuPatch)
 		} else if str, ok := spec.(map[string]interface{})["linkType"].(string); ok && strings.ToLower(str) == "ib" {
-			glog.V(2).Infof("defaultSriovNetworkNodePolicy(): set default mtu to 2048 for %v", name)
+			glog.V(2).Infof("mutateSriovNetworkNodePolicy(): set default mtu to 2048 for %v", name)
 			patchs = append(patchs, defaultIBMtuPatch)
 		}
 	}
+	// Device with InfiniBand link type requires isRdma to be true
+	if str, ok := spec.(map[string]interface{})["linkType"].(string); ok && strings.ToLower(str) == "ib" {
```

Comment:
what's the concern of not checking ?

Reply:
If not checking, the webhook will mutate the `isRdma` regardless of whether it is set by the user or not. I think it shall only set the default value when the user doesn't set it. And if the user sets an incorrect value, the webhook shall block it during validation and give an error message.

Reply:
If `isRdma` is not set by user, it will by default be set to false.
Since the only value user should set is true when using InfiniBnd link type, we don't want user to bother setting this argument, instead always set to true. This is different from the validation of vfio-pci deviceType && isRdma, because setting vfio-pci deviceType doesn't mean isRdma is true.

Reply:
What if a user defines a policy with IB device and `isRdma` false explicitly. I think we shall tell the user that it is an invalid configuration, instead of mutating it silently.

Reply:
If user defines a policy with conflicted settings, operator webhook will mutate it silently. There is also a check in the validation webhook in this PR, but mutating comes first then validation.
What we can do is to document the behavior of Operator webhook along with isRdma flag.Something like isRdma will be automatically configured when InfiniBand link type is specified.

I was thinking to add a warning message in Operator webhook that could be passed to API client, but this feature is not yet available in released api version. Once it's ready, we can use the warning message to indicate the automatic mutation.

Reply:
This was part of the discussion we had before about reducing the complex of API options that user may need to set.
When we added isRdma flag in API initially, it was for the purpose of supporting RoCE, but for now, it looks useless for InfiniBand use case and it doesn't make sense for user to specify it as we only allow one fixed value. I'm trying to hide this option as much as possible for future cases which doesn't need a explicit setting.